### PR TITLE
Me 68

### PR DESCRIPTION
**Pull request by dhiraj-toptal**

### CODE REVIEW

File: Controller/Webhook/Transaction/Confirmed.php [hunk 1]
```diff
@@ -60,127 +63,57 @@ class Confirmed extends Action implements CsrfAwareActionInterface
...
      *
-     * @return \Magento\Framework\App\ResponseInterface|\Magento\Framework\Controller\Result\Json|ResultInterface
+     * @return \Magento\Framework\App\ResponseInterface|ResultInterface
      * @throws \Magento\Framework\Exception\NoSuchEntityException
      */
     public function execute()
     {
-        sleep(15);// phpcs:ignore
         if (!$this->balancepayConfig->isActive()) {
             return $this->resultFactory->create(ResultFactory::TYPE_FORWARD)->forward('noroute');
```
Comment (ynaxonBalance) [inline Controller/Webhook/Transaction/Confirmed.php:112]: What does it do?

Comment (ynaxonBalance) [reply Controller/Webhook/Transaction/Confirmed.php:112]: In general ,shouldn't we unsubscribe to webhooks when balance payment is off? (and then subscribe again when it's on)

File: Controller/Webhook/Transaction/Confirmed.php [hunk 2]
```diff
@@ -60,127 +63,57 @@ class Confirmed extends Action implements CsrfAwareActionInterface
...
-            );
-        }
-        return $this;
+        $content = $this->getRequest()->getContent();
+        $headers = $this->getRequest()->getHeaders()->toArray();
+        $this->balancepayConfig->log('Webhook\Checkout\Confirmed::execute() ', 'debug', [
+            'content' => $content,
+            'headers' => $headers,
+        ]);
+        $this->helperData->processWebhook($content, $headers, self::WEBHOOK_CONFIRMED_NAME);
```
Comment (ynaxonBalance) [inline Controller/Webhook/Transaction/Confirmed.php]: I think the logic for processing a webhook should be defined in its own class (like `TransactionConfirmedWebhookProcessor`)

File: Helper/Data.php [hunk 3]
```diff
@@ -130,6 +176,213 @@ public function getSellerIdByProductId($productId = '')
...
+            $resBody = [
+                "error" => 1,
+                "message" => $e->getMessage(),
+            ];
+            if ($this->balancepayConfig->isDebugEnabled()) {
+                $resBody["trace"] = $e->getTraceAsString();
+            }
+        }
+
+        return $this->jsonResultFactory->create()
```
Comment (ynaxonBalance) [inline Helper/Data.php]: Why do you return http response here?

File: Helper/Data.php [hunk 4]
```diff
@@ -130,6 +176,213 @@ public function getSellerIdByProductId($productId = '')
         return $sellerId;
     }
 
+    /**
+     * @param $content
+     * @param $headers
+     * @return \Magento\Framework\Controller\Result\Json
+     * @throws \Magento\Framework\Exception\NoSuchEntityException
+     */
+    public function processWebhook($content, $headers, $webhookName)
```
Comment (ynaxonBalance) [inline Helper/Data.php]: move to a separate class that is responsible for handling webhooks from Balance, like `BalanceWebhookManager`

File: Helper/Data.php [hunk 5]
```diff
@@ -130,6 +176,213 @@ public function getSellerIdByProductId($productId = '')
...
+                $resBody["trace"] = $e->getTraceAsString();
+            }
+        }
+
+        return $this->jsonResultFactory->create()
+            ->setHttpResponseCode(Response::HTTP_OK)
+            ->setData($resBody);
+    }
+
+    public function addWebhookQue($params, $name)
```
Comment (ynaxonBalance) [inline Helper/Data.php]: move to a separate class that is responsible for handling webhooks from Balance, like `BalanceWebhookManager`

Comment (ynaxonBalance) [reply Helper/Data.php]: Fix method name: `enqueueWebhook`

File: Helper/Data.php [hunk 6]
```diff
@@ -130,6 +176,213 @@ public function getSellerIdByProductId($productId = '')
...
+            'attempts' => 1
+        ]);
+        $webhookModel->save();
+    }
+
+    /**
+     * @param $params
+     * @param $order
+     */
+    public function processConfirmedWebhook($params, $order)
```
Comment (ynaxonBalance) [inline Helper/Data.php]: Move to a class that is responsible from proessing `transaction/confirmed` webhook, like `TransactionConfirmedWebhookProcessor`

File: Helper/Data.php [hunk 7]
```diff
@@ -130,6 +176,213 @@ public function getSellerIdByProductId($productId = '')
...
+        }
+    }
+
+    /**
+     * @param $params
+     * @param $order
+     * @return void
+     * @throws \Magento\Framework\Exception\NoSuchEntityException
+     */
+    public function processChargedWebhook($params, $order)
```
Comment (ynaxonBalance) [inline Helper/Data.php]: Move to a class that is responsible from proessing transaction/confirmed webhook, like `TransactionChargedWebhookProcessor`

File: Helper/Data.php [hunk 8]
```diff
@@ -130,6 +176,213 @@ public function getSellerIdByProductId($productId = '')
...
+        try {
+            $chargeId = (string)$params['chargeId'];
+            $amount = (float)$params['amount'];
+            $orderPayment = $order->getPayment();
+
+            if (\strpos($orderPayment
+                    ->getAdditionalInformation(BalancepayMethod::BALANCEPAY_CHARGE_ID), $chargeId) === false) {
+                if (!$orderPayment
+                        ->getAdditionalInformation(BalancepayMethod::BALANCEPAY_IS_AUTH_CHECKOUT)
+                    && round((float)$order->getBaseGrandTotal()) !== round($amount)) {
```
Comment (ynaxonBalance) [inline Helper/Data.php]: extract condition statements to variables and merge two conditions into one

File: Helper/Data.php [hunk 9]
```diff
@@ -130,6 +176,213 @@ public function getSellerIdByProductId($productId = '')
...
+                    $orderPayment->setIsFraudDetected(true)->save();
+                    $order->setStatus(Order::STATUS_FRAUD)->save();
+                    throw new LocalizedException(new Phrase("The charged amount doesn't match the order total!"));
+                }
+
+                $orderPayment
+                    ->setTransactionId($orderPayment
+                        ->getAdditionalInformation(BalancepayMethod::BALANCEPAY_CHECKOUT_TRANSACTION_ID))
```
Comment (ynaxonBalance) [inline Helper/Data.php]: redundant? you the the value of `$orderPayment`'s id to `$orderPayment`'s id

File: Helper/Data.php [hunk 10]
```diff
@@ -130,6 +176,213 @@ public function getSellerIdByProductId($productId = '')
...
+                    ->setIsTransactionPending(false)
+                    ->setIsTransactionClosed(true)
+                    ->setAdditionalInformation(
+                        BalancepayMethod::BALANCEPAY_CHARGE_ID,
+                        $orderPayment->getAdditionalInformation(
+                            BalancepayMethod::BALANCEPAY_CHARGE_ID,
+                            $chargeId
+                        ) . " \n" . $chargeId
```
Comment (ynaxonBalance) [inline Helper/Data.php]: What's the purpose here?

File: Helper/Data.php [hunk 11]
```diff
@@ -130,6 +176,213 @@ public function getSellerIdByProductId($productId = '')
...
+                    );
+
+                if (!$orderPayment
+                    ->getAdditionalInformation(BalancepayMethod::BALANCEPAY_IS_AUTH_CHECKOUT)) {
```
Comment (ynaxonBalance) [inline Helper/Data.php]: extract to variable

File: Helper/Data.php [hunk 12]
```diff
@@ -130,6 +176,213 @@ public function getSellerIdByProductId($productId = '')
...
+                    $orderPayment->capture(null);
+                }
+                $orderPayment->save();
+                $order->save();
+                return true;
+            } elseif ($chargeId !== (string)$order->getPayment()
+                    ->getAdditionalInformation(BalancepayMethod::BALANCEPAY_CHARGE_ID)) {
+                throw new LocalizedException(new Phrase("Charge ID mismatch!"));
+            }
```
Comment (ynaxonBalance) [inline Helper/Data.php]: move this to be a negative condition before its related `if` statement to keep code cleaner:

```
$orderChargeId = (string)$order->getPayment()
                    ->getAdditionalInformation(BalancepayMethod::BALANCEPAY_CHARGE_ID);
if ($chargeid !== $orderChargeId) {
    throw new LocalizedException(new Phrase("Charge ID mismatch!"));
}

if (\strpos($orderPayment
                    ->getAdditionalInformation(BalancepayMethod::BALANCEPAY_CHARGE_ID), $chargeId) === false) {
                if (!$orderPayment
                        ->getAdditionalInformation(BalancepayMethod::BALANCEPAY_IS_AUTH_CHECKOUT)
                    && round((float)$order->getBaseGrandTotal()) !== round($amount)) {
    // ...
}
```

File: Helper/Data.php [hunk 13]
```diff
@@ -130,6 +176,213 @@ public function getSellerIdByProductId($productId = '')
...
+            return false;
+        }
+    }
+
+    /**
+     * @param $content
+     * @return array
+     * @throws LocalizedException
+     */
+    public function validateSignature($content, $headers, $webhookName): array
```
Comment (ynaxonBalance) [inline Helper/Data.php]: move to a separate class that is responsible for handling webhooks from Balance, like `BalanceWebhookManager`

File: Helper/Data.php [hunk 14]
```diff
@@ -130,6 +176,213 @@ public function getSellerIdByProductId($productId = '')
...
+        return $params;
+    }
+
+    /**
+     * @param $content
+     * @param $headers
+     * @return \Magento\Framework\Controller\Result\Json
+     * @throws \Magento\Framework\Exception\NoSuchEntityException
+     */
+    public function checkoutProcess($content, $headers)
```
Comment (ynaxonBalance) [inline Helper/Data.php]: is this a webhook listener? if so, for which webhook?

File: Helper/Data.php [hunk 15]
```diff
@@ -210,4 +487,29 @@ public function isCustomerGroupAllowed()
...
+    public function processWebhookCron($params, $webhook)
+    {
+        $isTransactionSuccess = false;
+        $order = $this->orderFactory->create()->loadByIncrementId((string)$params['externalReferenceId']);
+        if (!$order || !$order->getId()) {
+            $attempts = $webhook->getAttempts() + 1;
+            $this->updateAttempts($webhook->getEntityId(), 'attempts', $attempts);
+            throw new LocalizedException(new Phrase("No matching order!"));
+        }
+        if ($webhook->getName() == Confirmed::WEBHOOK_CONFIRMED_NAME) {
```
Comment (ynaxonBalance) [inline Helper/Data.php]: Where is this enum defined? I think there should be one enum `BalanceWebhook`, that should include `TransactionConfirmed` and `TransactionCharged` (and any other webhook in the future

File: Controller/Webhook/Transaction/Confirmed.php (hunk 1 above)
Comment (dhiraj-toptal) [reply Controller/Webhook/Transaction/Confirmed.php:112]: This we manage in the file - Balancepay/Balancepay/Observer/Config/Save.php

File: Helper/Data.php (hunk 3 above)
Comment (dhiraj-toptal) [reply Helper/Data.php]: This response will be returned to balance from where the webhook is called.

File: Helper/Data.php (hunk 10 above)
Comment (dhiraj-toptal) [reply Helper/Data.php]: To set the charge id in the payment information.

File: Helper/Data.php (hunk 12 above)
Comment (dhiraj-toptal) [reply Helper/Data.php]: This we cannot move up as charge id is not yet set to the payment object.

File: Helper/Data.php (hunk 14 above)
Comment (dhiraj-toptal) [reply Helper/Data.php]: Created a new class to handle this.

File: Helper/Data.php (hunk 15 above)
Comment (dhiraj-toptal) [reply Helper/Data.php]: We have defined a name of the webhook in the webhook_cron table.

**Comment by dhiraj-toptal** [comment]
HI @ynaxonBalance 
Can you review the PR?

**Review by ynaxonBalance — APPROVED** [review]

**Pull request merged**